cantus_text_syllabification.py

```diff
@@ -0,0 +1,206 @@
+STR_BEGINS_W_HYPHEN_REGEX = re.compile(r"^\-")
+# Matches a string that ends with a hyphen
+STR_ENDS_W_HYPHEN_REGEX = re.compile(r"\-$")
+# Matches pipes and missing music sectioners ("{" and "}")
+TEXT_SECTIONER_REGEX = re.compile(r"(\|)|(\{.*?\})(?!\s*?\{)")
+
+
+def _clean_text(test: str) -> str:
+    """
+    Removes invalid characters from the text string.
+
+    test [str]: string to clean
+
+    returns [str]: cleaned string
+    """
+    return INVALID_CHAR_REGEX.sub("", test)
+
+
+def _detect_invalid_characters(text: str) -> bool:
+    """
+    Detects invalid characters in the text string.
+
+    text [str]: string to check
+
+    returns [bool]: True if invalid characters are present, False otherwise
+    """
+    return bool(INVALID_CHAR_REGEX.search(text))
+
+
+def _prepare_string_for_syllabification(word_str: str) -> "tuple[str, bool, bool]":
+    """
+    Complete preparation of a string before syllabification.
+    All letters are converted to lowercase. Hyphens are removed
+    from the beginning and end of the string, and the presence
+    of these hyphens is recorded.
+
+    word_str [str]: string to prepare
+
+    returns [tuple[str,bool,bool]]: prepared string, whether a hyphen
+        was removed from the beginning of the string, whether a hyphen
+        was removed from the end of the string
+    """
+    word_str, start_hyphen = STR_BEGINS_W_HYPHEN_REGEX.subn("", word_str)
+    word_str, end_hyphen = STR_ENDS_W_HYPHEN_REGEX.subn("", word_str)
+    return word_str, bool(start_hyphen), bool(end_hyphen)
+
+
+def _split_text_sections(text: str) -> "list[str]":
+    """
+    Splits a text string into sections based on the presence of
+    curly braces "{}", square brackets "[]",
+    and pipes "|". These special characters
+    are captured in the sections.
+
+    text [str]: text to split
+
+    returns [list[str]]: list of text sections
+    """
+    text_sections = TEXT_SECTIONER_REGEX.split(text)
+    # Remove extra spaces from sections and remove empty sections/None-type sections
+    # caused by section split.
+    text_sections = [
+        section.strip() if section is not None else "" for section in text_sections
+    ]
+    text_sections = list(filter(lambda x: x != "", text_sections))
+    return text_sections
+
+
+def syllabify_text(
+    text: str,
+    clean_text: bool = False,
+    flatten_result: bool = False,
+) -> Union["list[list[str]]", str]:
+    """
+    Syllabifies a text string that has been encoded in the style
+    of the Cantus Database. Texts are syllabified word by word,
+    except in special cases outlined in the README.
+
+    text [str]: text to syllabify
+    clean_text [bool]: if True, removes invalid characters from
+        the text string; if False, raises a ValueError if invalid
+        characters are detected
+    flatten_result [bool]: if True, returns a string of the
+        syllabified text instead of a list of lists with hyphens separating
```

Comment:
could be clarified:
`syllabified text (instead of a list of lists) with hyphens separating` - at first, it reads like the hyphens are somehow separating the lists